File: python/techai_webutils/src/techai_webutils/core/interfaces/bulkhead.py
"""Bulkhead interface for concurrency limiting.

Mirrors Go's ``interfaces.Bulkhead``.
"""

from __future__ import annotations

from abc import ABC, abstractmethod
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from collections.abc import Awaitable, Callable


class Bulkhead(ABC):
    """Limits concurrent access to a resource."""

    @abstractmethod
    async def execute(self, fn: Callable[[], Awaitable[object]]) -> object:
        """Run fn within the concurrency limit.

        Blocks until a slot is available.
        """
        ...

    @abstractmethod
    async def try_execute(self, fn: Callable[[], Awaitable[object]]) -> object:
        """Attempt to run fn without blocking.

        Raises ``BulkheadFullError`` if no slots are available.
        """
        ...

File: go/core/interfaces/bulkhead.go
package interfaces

import "context"

// Bulkhead limits concurrent access to a resource, preventing unbounded
// parallelism from exhausting goroutines, file descriptors, and memory.
//
// Implementations: channel semaphore (default).
//
// All service code depends on this interface, never on a concrete
// implementation. Swap implementations via the bulkhead.New factory.
type Bulkhead interface {
	// Execute runs fn within the concurrency limit. Blocks until a slot is
	// available or ctx is cancelled.
	Execute(ctx context.Context, fn func() error) error

	// TryExecute attempts to run fn without blocking. Returns an error if no
	// slots are available.
	TryExecute(fn func() error) error
}
